File: backend/src/infrastructure/persistence/models/srg.py
from django.conf import settings
from django.db import models

from infrastructure.persistence.models.user import Concesionaria
from .base import BaseModel


class SrgType(models.TextChoices):
    WARRANTY = "WARRANTY", "Garantía"
    CAMPAIGN = "CAMPAIGN", "Campaña"


class SrgStatus(models.TextChoices):
    PROCESO = "PROCESO", "Proceso"
    PENDIENTE = "PENDIENTE", "Pendiente"
    PREAPROBADO = "PREAPROBADO", "Preaprobado"
    APROBADO = "APROBADO", "Aprobado"
    RETORNADO = "RETORNADO", "Retornado"
    NEGADO = "NEGADO", "Negado"


class Srg(BaseModel):
    ot = models.CharField(max_length=20, unique=True, db_index=True)
    srg_type = models.CharField(max_length=10, choices=SrgType.choices)
    status = models.CharField(max_length=15, choices=SrgStatus.choices, default=SrgStatus.PROCESO)
    concesionaria = models.CharField(max_length=20, choices=Concesionaria.choices, db_index=True)
    asesor = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.PROTECT,
        related_name="srgs",
    )

    # Common header fields
    vin = models.CharField(max_length=17, db_index=True)
    placa = models.CharField(max_length=10, db_index=True)
    vehicle_model = models.CharField(max_length=50)
    vehicle_color = models.CharField(max_length=50, blank=True)
    vehicle_year = models.PositiveSmallIntegerField()
    km_apertura = models.PositiveIntegerField()
    sede = models.CharField(max_length=20, choices=Concesionaria.choices)

    # Warranty-specific
    nro_garantia = models.CharField(max_length=20, blank=True)
    warranty_type_code = models.CharField(max_length=10, blank=True)
    warranty_type_name = models.CharField(max_length=200, blank=True)

    # Campaign-specific
    campaign_code = models.CharField(max_length=20, blank=True)

    # Status timestamps
    fecha_envio_marca = models.DateTimeField(null=True, blank=True)
    fecha_aprobacion = models.DateField(null=True, blank=True)

    class Meta:
        db_table = "srgs"
        ordering = ["-created_at"]
        indexes = [
            models.Index(fields=["concesionaria", "status"]),
            models.Index(fields=["concesionaria", "srg_type"]),
        ]

    def __str__(self) -> str:
        return f"OT {self.ot} [{self.srg_type}] — {self.status}"

File: frontend/src/types/srg.ts
export type SrgType   = "WARRANTY" | "CAMPAIGN";
export type SrgStatus = "PROCESO" | "PENDIENTE" | "PREAPROBADO" | "APROBADO" | "RETORNADO" | "NEGADO";

export interface Srg {
  id:             string;
  ot:             string;
  srg_type:       SrgType;
  status:         SrgStatus;
  concesionaria:  string;
  asesor_id:      string;
  vin:            string;
  placa:          string;
  vehicle_model:  string;
  vehicle_color:  string;
  vehicle_year:   number;
  km_apertura:    number;
  sede:           string;
  // Warranty
  nro_garantia?:      string | null;
  warranty_type_code?: string | null;
  warranty_type_name?: string | null;
  // Campaign
  campaign_code?:      string | null;
  // Dates
  fecha_envio_marca?:  string | null;
  fecha_aprobacion?:   string | null;
  created_at:          string;
}

export interface SrgPart {
  id:             string;
  catalog_code:   string;
  name_es:        string;
  quantity:       number;
  unit_price:     string;
  part_origin:    string;
  invoice_number: string;
  created_at:     string;
}

export type ReceptionState = "PENDIENTE" | "PARCIAL" | "COMPLETA";
export type ReturnState    = "NO_APLICA" | "PENDIENTE" | "PARCIAL" | "COMPLETA";
export type SrgEventType =
  | "PART_REQUESTED"
  | "RECEPTION_REGISTERED"
  | "WORK_REGISTERED"
  | "CORE_RETURN_DECLARED"
  | "RETURN_CONFIRMED"
  | "STATUS_CHANGED";

export type MovementType = Exclude<SrgEventType, "PART_REQUESTED" | "STATUS_CHANGED">;

export interface PartLedger {
  id:                 string;
  catalog_code:       string;
  name_es:            string;
  unit_price:         string;
  part_origin:        string;
  invoice_number:     string;
  created_at:         string;
  requested:          number;
  received:           number;
  used:               number;
  returned_declared:  number;
  returned_confirmed: number;
  pending_reception:  number;
  pending_return:     number;
  reception_state:    ReceptionState;
  return_state:       ReturnState;
  closed:             boolean;
}

export interface SrgEvent {
  id:          string;
  srg_id:      string;
  srg_part_id: string | null;
  actor_id:    string | null;
  actor_role:  string;
  actor_label: string;
  part_label:  string;
  event_type:  SrgEventType;
  quantity:    number | null;
  state_from:  string;
  state_to:    string;
  note:        string;
  location:    string;
  created_at:  string;
}

export interface CampaignBody {
  id:          string;
  srg_id:      string;
  update_name: string;
  image_link:  string;
  modified_by: string;
  updated_at:  string;
}

export interface CatalogParam {
  id:           string;
  param_type:   string;
  code:         string;
  name:         string;
  concesionaria: string;
}

// ── Status transition helpers ──────────────────────────────────────────────
export const STATUS_LABELS: Record<SrgStatus, string> = {
  PROCESO:     "Proceso",
  PENDIENTE:   "Pendiente",
  PREAPROBADO: "Preaprobado",
  APROBADO:    "Aprobado",
  RETORNADO:   "Retornado",
  NEGADO:      "Negado",
};

export const SRG_TYPE_LABELS: Record<SrgType, string> = {
  WARRANTY: "Garantía",
  CAMPAIGN: "Campaña",
};
